modello-core/src/main/java/org/codehaus/modello/plugin/AbstractModelloGenerator.java

```diff
@@ -198,11 +208,14 @@ public static String singular(String name) {
             return name;
         }
 
-        if (name.endsWith("ies")) {
+        if (PLURAL_EXCEPTION.containsKey(name)) {
+            return PLURAL_EXCEPTION.get(name);
+        } else if (name.endsWith("ies")) {
             return name.substring(0, name.length() - 3) + "y";
-        } else if (name.endsWith("es") && name.endsWith("ches")) {
-            return name.substring(0, name.length() - 2);
-        } else if (name.endsWith("xes")) {
+        } else if (name.endsWith("zzes")) {
+            return name.substring(0, name.length() - 3);
+        } else if (name.endsWith("ches") || name.endsWith("xes") || name.endsWith("ses") || name.endsWith("oes")
```

Comment:
The `oes` rule is a bit more complicated.
It usually looses the final `s`, but there are exceptions such as heroes, potatoes, tomatoes, echoes, torpedoes, vetoes, cargoes, haloes, and mosquitoes.
I think the rule should be loose the final `s`, with exceptions such as:
```
    "heroes", "hero",
    "potatoes", "potato",
    "tomatoes", "tomato",
    "echoes", "echo",
    "vetoes", "veto",
    "torpedoes", "torpedo",
    "cargoes", "cargo",
    "haloes", "halo",
    "mosquitoes", "mosquito",
    "buffaloes", "buffalo"
```